### DISCUSSION TITLE
Difference between `std::mem_size_of` and `::SHADER_SIZE`

### DISCUSSION THREAD

**Discussion by davids91** [gfx-rs/wgpu#5947] [Q&A] [answered]
Anyone would know why is there a difference between std::mem::size_of and ::SHADER_SIZE ( from encase ) ? 
For the below struct: 
```
#[derive(Clone, ShaderType)]
pub struct MyData {
    pub a: Vec3,
    pub b: Vec3,
    pub(crate) s: u32,
    pub(crate) dim: u32,
}
```

this fails: 
```
assert_eq!(
    std::mem::size_of::<MyData>(),
    MyData::SHADER_SIZE.get() as usize
);
```

with: 
```
assertion `left == right` failed
  left: 32
 right: 48
```

I tried to calculate the size in wgpu for that struct but I think I messed sth up, because it should be fine...
alignOf(Vec3) should be 16, but upon reordering the struct to: 
```
#[derive(Clone, ShaderType)]
pub struct MyData {
    pub a: Vec3,
    pub(crate) s: u32,
    pub b: Vec3,
    pub(crate) dim: u32,
}
```

I get the same result

Why might that be and how could one fix it?

**Comment by kpreid** [comment, marked as answer]
You haven't put `#[repr(C)]` on your struct, so reordering the fields isn't guaranteed to do anything to its layout.

If you're planning to use `encase`, why are you worrying about the layout of your Rust structs anyway? The way `encase` works (if I understand it correctly) is that it _copies the fields of_ your structs into the buffers it manages.

**Reply by davids91** [reply]
Because I did not know this aspect of `encase` yet, thank you for the info! 

So if I understand correctly it is irrelevant that `SHADER_SIZE` and `std::mem` size is different.

**Reply by teoxoy** [reply]
Yeah, the size difference is to be expected since the layout of WGSL structs is different.

**Reply by davids91** [reply]
Thank you :)
